resourceIndexer/inventory.js

```diff
@@ -0,0 +1,157 @@
+  return inventory;
+}
+
+
+/**
+ * Returns a list of objects in the requested S3 inventory, which itself represents a list of
+ * objects + versions within a specific bucket+prefix.
+ * 
+ * Keys which match a delete marker key are considered deleted and are removed here.
+ * This includes all versions past before and after the delete marker. Internally we haven't
+ * explicitly considered what it means to "delete" a resource, so this is the most conservative
+ * implementation.
+ * 
+ * The option exists to source the inventory files locally (useful for dev purposes to avoid constant
+ * downloads from S3)
+ */
+export const parseInventory = async ({bucket, prefix, name, local}) => {
+  let objects;
+  try {
+    if (local) {
+      objects = await fetchInventoryLocal({bucket, prefix, name});
+    } else {
+      objects = await fetchInventoryRemote({bucket, prefix, name});
+    }
+  } catch (e) {
+    logger.error(`Error while fetching s3 inventory for ${name}: ${e.message}`)
+    return [];
+  }
+  /* Scan the objects to find delete markers and filter out both the markers
+  themselves and all other keys which are identical to a delete marker. We may
+  want to relax this in the future, e.g. if a v1 dataset was deleted but a v2
+  dataset exists (and the resulting nextstrain.org URLs would be the same) then
+  we may want to consider the "deleted" objects. */
+  const deletedKeys = new Set();
+  objects = objects.filter((item) => {
+    if (item.IsDeleteMarker === "true") {
+      deletedKeys.add(item.Key);
+      return false;
+    }
+    return true;
+  }).map((item) => {
+    if (deletedKeys.has(item.Key)) {
+      return false;
+    }
+    // a bucket without versioning enabled will not have the `IsLatest` key
+    if (!Object.hasOwn(item, 'IsLatest')) item.IsLatest = "true";
+    // There are some keys in versioned buckets where the earliest version
+    // doesn't have a version ID. I'm not sure what these represent - e.g. see
+    // s3://nextstrain-data/WNV_NA_tree.json from 2018-05-09. I think it's safe
+    // to filter them out here.
+    if (item.VersionId === '') return false;
+    return item;
+  }).filter((item) => !!item) // filter out empty items from above
+
+  /* rename / prune / add properties as I find the default S3 properties /
+  values awkward to work with */
+  return objects.map((item) => {
+    return {
+      timestamp: DateTime.fromISO(item.LastModifiedDate),
+      date: item.LastModifiedDate.split("T")[0],
+      key: item.Key,
+      versionId: item.VersionId, // will be undefined if bucket is not versioned
+      latest: item.IsLatest==='true',
+    }
+  });
+}
+
+/**
+ * Parses a S3 inventory manifest JSON file
+ * @param {object} manifest
+ * @returns {object} object.schema = string[]
+ *                   object.inventoryKey = string
+*/
+function _parseManifest(manifest) {
+  return {
+    schema: manifest.fileSchema.split(",").map((f) => f.trim()),
+    inventoryKey: manifest.files[0].key
```

Comment:
Yeah, perhaps. The docs aren't clear. Would a fatal error be acceptable to you in this case?

Reply:
Since I can't find an example of this I've elected to throw a fatal error with a helpful error message. I doubt we'll ever encounter this but I agree that if we do we shouldn't silently take the first.

Reply:
Ok, sure.

Reply:
Just seeing this review after the fatal error was surfaced in #975. Thanks for the sharp eyes @tsibley and helpful error message @jameshadfield!